tests/test_main.py

```diff
@@ -2986,17 +2988,27 @@ class Bar:
     assert ta.validate_python(bar) is bar
 
 
-def test_deferred_core_schema() -> None:
-    class Foo(BaseModel):
-        x: 'Bar'
+@pytest.mark.parametrize('is_dataclass', [False, True])
+def test_deferred_core_schema(is_dataclass: bool) -> None:
+    if is_dataclass:
+
+        @pydantic_dataclass
+        class Foo:
+            x: 'Bar'
+    else:
+
+        class Foo(BaseModel):
+            x: 'Bar'
 
+    assert isinstance(Foo.__pydantic_core_schema__, MockCoreSchema)
     with pytest.raises(PydanticUserError, match='`Foo` is not fully defined'):
-        Foo.__pydantic_core_schema__
+        Foo.__pydantic_core_schema__['type']
```

Comment:
Why is the `['type']` addition here necessary?

Reply:
I think you have to access something on it to get it to produce the error now, not just accessing the value itself

Reply:
Yep exactly. It works similarly as other `MockValSer`s where the model building happens on the attribute accessing